cpp/src/arrow/array/builder_adaptive.h

```diff
@@ -83,6 +99,9 @@ class ARROW_EXPORT AdaptiveUIntBuilder : public internal::AdaptiveIntBuilderBase
     pending_valid_[pending_pos_] = 1;
     ++pending_pos_;
 
+    if (val & expand_int_size_mask_) {
```

Comment:
Add `ARROW_PREDICT_FALSE`. This can happen at most 4 times over the lifetime of a builder.